xarray/core/indexing.py

```diff
@@ -1288,6 +1290,47 @@ def __init__(self, array):
         self.array = array
 
 
+class ArrayApiIndexingAdapter(ExplicitlyIndexedNDArrayMixin):
+    """Wrap an array API array to use explicit indexing."""
+
+    __slots__ = ("array",)
+
+    def __init__(self, array):
+        if not hasattr(array, "__array_namespace__"):
+            raise TypeError(
+                "ArrayApiIndexingAdapter must wrap an object that "
+                "implements the __array_namespace__ protocol"
+            )
+        self.array = array
+
+    def __getitem__(self, key):
+        if isinstance(key, BasicIndexer):
+            return self.array[key.tuple]
+        elif isinstance(key, OuterIndexer):
+            # manual orthogonal indexing (implemented like DaskIndexingAdapter)
+            key = key.tuple
+            value = self.array
+            for axis, subkey in reversed(list(enumerate(key))):
+                value = value[(slice(None),) * axis + (subkey, Ellipsis)]
+            return value
+        else:
+            assert isinstance(key, VectorizedIndexer)
```

Comment:
nit pick:  since `assert` statements are removed when  python is invoked with [`-O` and `-OO` parameters,](https://docs.python.org/3/using/cmdline.html#cmdoption-o) could we  raise a proper exception?

Reply:
Yes we should raise I think

Reply:
Replaced with a `TypeError`